CromwellRefdiskManifestCreator/src/main/java/org/broadinstitute/manifestcreator/CromwellRefdiskManifestCreatorApp.java

```diff
@@ -0,0 +1,151 @@
+import java.util.List;
+import java.util.Stack;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+import java.util.zip.CRC32;
+
+public class CromwellRefdiskManifestCreatorApp {
+
+  private static final Logger logger = LogManager.getLogger(CromwellRefdiskManifestCreatorApp.class);
+
+  public static void main(String[] args) throws IOException, InterruptedException {
+    Configurator.setRootLevel(Level.DEBUG);
+
+    if (args.length < 4) {
+      logger.error("Wrong number of parameters.");
+      logger.error("Usage: java -jar cromwell-refdisk-manifest-creator-app.jar <number of parallel threads> " +
+              "<image_name> <directory_path_to_scan> <output file path");
+      System.exit(1);
+    }
+
+    String directoryToScan = args[2];
+    File rootDirectory = new File(directoryToScan);
+    if (!rootDirectory.exists() || !rootDirectory.isDirectory()) {
+      logger.error("Root directory " + directoryToScan + " doesn't exist.");
+      System.exit(1);
+    }
+
+    int totalNumberOfFiles = countFiles(rootDirectory);
+    logger.info("Parsing " + totalNumberOfFiles + " files");
+
+    ReferenceDiskManifest manifest = new ReferenceDiskManifest();
+    String imageName = args[1];
+    manifest.setImageIdentifier(imageName);
+
+    int nThreads = Integer.parseInt(args[0]);
+    CountDownLatch countDownLatch = new CountDownLatch(nThreads);
+    ExecutorService executorService = Executors.newFixedThreadPool(nThreads);
+
+    Path rootPath = rootDirectory.toPath();
+    Stack<File> stack = new Stack<>();
+    stack.push(rootDirectory);
+    List<ReferenceFile> referenceFiles = Collections.synchronizedList(new ArrayList<>());
+    for (int i = 0; i < nThreads; i++) {
+      executorService.submit(() -> {
+        try {
+          doWork(rootPath, referenceFiles, stack, totalNumberOfFiles, countDownLatch);
+        } catch (CRC32Exception e) {
+          logger.error("Error occurred. Shutting down.", e);
+          executorService.shutdownNow();
+          System.exit(1);
+        }
+      });
+    }
+
+    countDownLatch.await();
+
+    String manifestFilePath = args[3];
+    File manifestFile = new File(manifestFilePath);
+    if (manifestFile.exists() || !manifestFile.createNewFile()) {
+      logger.error("File " + manifestFile.getAbsolutePath() + " already exists or cannot be created.");
+    } else {
+      referenceFiles.forEach(referenceFile -> manifest.getFiles().add(referenceFile));
+      new ObjectMapper().writerWithDefaultPrettyPrinter().writeValue(manifestFile, manifest);
+      logger.info("Finished. Total number of reference files: " + manifest.getFiles().size());
+    }
+
+    executorService.shutdown();
+  }
+
+  private static void doWork(Path rootPath, List<ReferenceFile> accumulator, Stack<File> stack, int totalNumberOfFiles, CountDownLatch countDownLatch) throws CRC32Exception {
```

Comment:
It seems a little odd to me that you go to the effort of pre-traversing the directory to find all of the files inside it, but instead of using that list of files you found to seed the `stack` variable, you instead throw away all the files that you found and only keep the count - and the `doWork` method has to recalculate the stack as it goes along (leaving open the possibility that only one thread can be working even though there are 100s of files still waiting to be discovered)

Reply:
Sounds good, I'll fix that. I added multithreading in the very end, after implementing all the main logic and finding out that it works very slow, so I didn't think this through.